parser.y

```diff
@@ -1268,12 +1271,50 @@ AlterTableSpec:
 			PartitionNames: $4.([]model.CIStr),
 		}
 	}
-|	"TRUNCATE" "PARTITION" PartitionNameList %prec lowerThanComma
+|	"TRUNCATE" "PARTITION" AllOrPartitionNameList
 	{
-		$$ = &ast.AlterTableSpec{
+		allOrPartitionNames := $3.(*ast.AllOrPartitionNames)
+		ret := &ast.AlterTableSpec{
 			Tp: ast.AlterTableTruncatePartition,
-			PartitionNames: $3.([]model.CIStr),
 		}
+		if allOrPartitionNames.All {
+			ret.OnAllPartitions = true
+			yylex.AppendError(yylex.Errorf("The TRUNCATE PARTITION ALL clause is parsed but ignored by all storage engines."))
+			parser.lastErrorAsWarn()
+		} else {
+			ret.PartitionNames = allOrPartitionNames.PartitionNames
+		}
+		$$ = ret
+	}
+|	"OPTIMIZE" "PARTITION" AllOrPartitionNameList
```

Comment:
`OPTIMIZE TABLE REPAIRE PARTITION` syntax supports `NO_WRITE_TO_BINLOG` option. See: [`OPTIMIZE PARTITION` definition in MySQL's parser file](https://github.com/mysql/mysql-server/blob/817fcb883a8457fcf507156bb1a2052e887acc4c/sql/sql_yacc.yy#L7896).

Reply:
NO_WRITE_TO_BINLOG PR: https://github.com/pingcap/parser/pull/456

Reply:
Thanks